contracts/compliance/modular/modules/InitialLockupPeriodModule.sol

```diff
@@ -0,0 +1,160 @@
+
+    struct LockedTokens {
+        uint256 amount;
+        uint256 releaseTimestamp;
+    }
+
+    mapping(address compliance => uint256 lockupPeriod) private _lockupPeriods;
+    mapping(address compliance => mapping(address user => LockedTokens[] lockedTokens)) private _lockedTokens;
+
+    /// @dev initializes the contract and sets the initial state.
+    function initialize() external initializer {
+        __AbstractModule_init();
+    }
+
+    /// @dev sets the lockup period for a compliance contract.
+    /// @param _lockupPeriod the lockup period in seconds.
+    function setLockupPeriod(uint256 _lockupPeriod) external onlyComplianceCall {
+        _lockupPeriods[msg.sender] = _lockupPeriod;
+
+        emit LockupPeriodSet(msg.sender, _lockupPeriod);
+    }
+
+    /// @inheritdoc IModule
+    function moduleTransferAction(address _from, address /*_to*/, uint256 _value) external override onlyComplianceCall { 
+        if (_from != address(0)) {
+            // Check if the user has enough unlocked tokens to transfer
+            uint256 availableAmount = _calculateAvailableAmount(msg.sender, _from);
+            if (_value > availableAmount) {
+                revert InsufficientBalanceTokensLocked(_from, _value, availableAmount);
+            }
+        }
+    }
+
+    /// @inheritdoc IModule
+    function moduleMintAction(address _to, uint256 _value) external override onlyComplianceCall {
+        _lockedTokens[msg.sender][_to].push(
+            LockedTokens({
+                amount: _value,
+                releaseTimestamp: block.timestamp + _lockupPeriods[msg.sender]
+            })
+        );
+    }
+
+    /// @inheritdoc IModule
+    // solhint-disable-next-line no-empty-blocks
+    function moduleBurnAction(address _from, uint256 _value) external override onlyComplianceCall { }
+
+    /// @inheritdoc IModule
+    function moduleCheck(address _from, address /*_to*/, uint256 _value, address _compliance) external 
+        view override returns (bool) {
+        return _from == address(0) || _calculateAvailableAmount(_compliance, _from) >= _value;
+    }
+
+    /// @inheritdoc IModule
+    function canComplianceBind(address /*_compliance*/) external view override returns (bool) {
+        return true;
+    }
+
+    /// @inheritdoc IModule
+    function isPlugAndPlay() external pure override returns (bool) {
+        return true;
+    }
+
+    /// @inheritdoc IModule
+    function name() external pure override returns (string memory _name) {
+        return "InitialLockupPeriodModule";
+    }
+
+    /// @dev calculates the available amount of unlocked tokens for a user.
+    /// @param _compliance the address of the compliance contract.
+    /// @param _user the address of the user.
+    /// @return _availableAmount the available amount of unlocked tokens.
+    function _calculateAvailableAmount(address _compliance, address _user) internal view returns (uint256 _availableAmount) {
```

Comment:
rework this function with the totalLocked value as explained above. It should process the array of locked tokens only if there's not enough "already free" tokens to fulfill the needs of the current transaction